statsmodels/tsa/statespace/regime_switching/kim_filter.py

```diff
@@ -0,0 +1,813 @@
+                filtered_prev_and_curr_regime_logprobs, axis=0)
+
+    def _regime_uncond_filtering(self, t,
+            filtered_prev_and_curr_regime_probs, state_batteries,
+            weighted_state_batteries, state_cov_batteries,
+            weighted_state_cov_batteries):
+
+        k_regimes = self._k_regimes
+
+        np.multiply(filtered_prev_and_curr_regime_probs.reshape(k_regimes,
+                k_regimes, 1), state_batteries, out=weighted_state_batteries)
+
+        np.sum(weighted_state_batteries, axis=(0, 1),
+                out=self._filtered_states[t, :])
+
+        np.multiply(filtered_prev_and_curr_regime_probs.reshape(k_regimes,
+                k_regimes, 1, 1), state_cov_batteries,
+                out=weighted_state_cov_batteries)
+
+        np.sum(weighted_state_cov_batteries, axis=(0, 1),
+                out=self._filtered_state_covs[t, :, :])
+
+    def _approximation_step(self, t, curr_regime,
+            filtered_prev_and_curr_regime_logprobs,
+            filtered_prev_cond_on_curr_regime_logprobs,
+            state_batteries, weighted_states, state_biases,
+            transposed_state_biases, state_bias_sqrs,
+            state_cov_batteries, state_covs_and_state_bias_sqrs,
+            weighted_state_covs_and_state_bias_sqrs, approx_state_cov):
+
+        k_states = self._k_states
+
+        curr_filter = self._kfilters[curr_regime]
+
+        approx_state = np.asarray(curr_filter.filtered_state[:, t])
+
+        if self._filtered_regime_logprobs[t, curr_regime] == -np.inf:
+
+            # Any value would be alright, since it is multiplied by zero weight
+            # in the next iteration
+            approx_state[:] = 0
+
+            approx_state_cov = \
+                    np.asarray(curr_filter.filtered_state_cov[:, :, t])
+
+            approx_state_cov[:, :] = 0
+
+            return
+
+        np.subtract(filtered_prev_and_curr_regime_logprobs[:, curr_regime],
+                self._filtered_regime_logprobs[t, curr_regime],
+                out=filtered_prev_cond_on_curr_regime_logprobs)
+
+        filtered_prev_cond_on_curr_regime_probs = \
+                filtered_prev_cond_on_curr_regime_logprobs
+
+        np.exp(filtered_prev_cond_on_curr_regime_logprobs,
+                out=filtered_prev_cond_on_curr_regime_probs)
+
+        np.multiply(filtered_prev_cond_on_curr_regime_probs.reshape(-1, 1),
+                state_batteries[:, curr_regime, :], out=weighted_states)
+
+        np.sum(weighted_states, axis=0, out=approx_state)
+
+        np.subtract(approx_state.reshape(1, -1, 1),
+                state_batteries[:, curr_regime, :].reshape(-1, k_states, 1),
+                out=state_biases)
+
+        np.subtract(approx_state.reshape(1, 1, -1),
+                state_batteries[:, curr_regime, :].reshape(-1, 1, k_states),
+                out=transposed_state_biases)
+
+        np.multiply(state_biases, transposed_state_biases, out=state_bias_sqrs)
```

Comment:
Aren't dot, multiply and outer all equivalent in this case? Or this is a question of style?